src/queues/mod.rs

```diff
@@ -0,0 +1,212 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, you can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::{
+    boxed::Box, error::Error, fmt::{self, Debug, Display, Formatter},
+};
+
+use futures::future::{self, Future};
+
+use self::notification::{Notification, NotificationType};
+pub use self::sqs::Queue as Sqs;
+use auth_db::{BounceSubtype, BounceType, Db, DbClient, DbError};
+use settings::Settings;
+
+mod mock;
+mod notification;
+pub mod sqs;
+#[cfg(test)]
+mod test;
+
+#[derive(Debug)]
+pub struct Queues {
+    bounce: Box<Incoming>,
+    complaint: Box<Incoming>,
+    delivery: Box<Incoming>,
+    notification: Box<Outgoing>,
+    db: DbClient,
+}
+
+pub trait Incoming: Debug + Sync {
+    fn receive(&'static self) -> ReceiveFuture;
+    fn delete(&'static self, message: Message) -> DeleteFuture;
+}
+
+type ReceiveFuture = Box<Future<Item = Vec<Message>, Error = QueueError>>;
+type DeleteFuture = Box<Future<Item = (), Error = QueueError>>;
+
+pub trait Outgoing: Debug + Sync {
+    fn send(&'static self, body: &Notification) -> SendFuture;
+}
+
+type SendFuture = Box<Future<Item = String, Error = QueueError>>;
+
+pub trait Factory {
+    fn new(id: String, settings: &Settings) -> Self;
+}
+
+#[derive(Debug, Default)]
+pub struct Message {
+    pub id: String,
+    pub notification: Notification,
+}
+
+#[derive(Debug)]
+pub struct QueueError {
+    description: String,
+}
+
+#[derive(Debug)]
+pub struct QueueIds {
+    pub bounce: String,
+    pub complaint: String,
+    pub delivery: String,
```

Comment:
Is it actually necessary for us to have three distinct SQS queues, one for each of bounces, complains, and delivery notifications?  Or is that a historical artifact of the way we added these in the auth-server, and they could actually all be delivered via the same queue if we configured things appropriately in AWS?

Reply:
I can't find anything in the docs that says you can't use the same queue for all three notification types, so I presume it would be fine. But @jrgm or @jbuck can probably confirm for us: Was it just our own preference to use separate queues for bounce, complaint and delivery notifications or are there objective reasons why it has to be that way?

Reply:
I think our own preference. It's possible to deliver to a single queue for all three types

Reply:
> I think our own preference. It's possible to deliver to a single queue for all three types

Thanks!

@rfk, I'm definitely in favour of simplifying it to a single incoming queue from a conceptual standpoint, I just wonder whether there is a benefit to keeping three queues for now from a pragmatic standpoint? I mean, if we did want to run this alongside/interchangeably with the existing auth server bounce-handling code at first, then wouldn't it make sense to do that while consuming from the exact same queues?

Reply:
Aye, continuing to use the three separate queues sounds for compatibility with auth-server sounds like the right place to start from, 👍